### PR TITLE
migrate `prefer_foreach` and `public_member_api_docs` to NodeLintRule (#1241)

### PR DESCRIPTION
**Pull request by pq**
Fixes: #1241.

/cc @scheglov @stereotype441

### CODE REVIEW

File: lib/src/rules/public_member_api_docs.dart [hunk 1]
```diff
@@ -76,7 +76,18 @@ class PublicMemberApiDocs extends LintRule {
...
+    registry.addConstructorDeclaration(this, visitor);
+    registry.addEnumConstantDeclaration(this, visitor);
+    registry.addEnumDeclaration(this, visitor);
+    registry.addFieldDeclaration(this, visitor);
+    registry.addFunctionTypeAlias(this, visitor);
+    registry.addTopLevelVariableDeclaration(this, visitor);
+  }
 }
 
 class _Visitor extends GeneralizingAstVisitor {
```
Comment (scheglov) [inline lib/src/rules/public_member_api_docs.dart]: Use `SimpleAstVisitor`?

Comment (pq) [reply lib/src/rules/public_member_api_docs.dart]: 👍  Done.

**Comment by pq** [comment]
Hardly conclusive but these relative benchmark timings are pretty compelling:


**Before:**

```
------------------------------------------------------
Timings                                             ms
------------------------------------------------------
public_member_api_docs                             195
avoid_private_typedef_functions                    166
invariant_booleans                                 151
prefer_foreach                                      84
avoid_positional_boolean_parameters                 73
avoid_renaming_method_parameters                    69
close_sinks                                         63
```

**After:**

```
------------------------------------------------------
Timings                                             ms
------------------------------------------------------
avoid_private_typedef_functions                    226
invariant_booleans                                 142
flutter_style_todos                                113
public_member_api_docs                             103
close_sinks                                         84
avoid_renaming_method_parameters                    83
avoid_positional_boolean_parameters                 82
```

**Comment by pq** [comment]
And after migrating to `SimpleAstVisitor`:

```
------------------------------------------------------
Timings                                             ms
------------------------------------------------------
avoid_private_typedef_functions                    174
invariant_booleans                                 118
public_member_api_docs                              70
avoid_positional_boolean_parameters                 69
close_sinks                                         64
flutter_style_todos                                 63
avoid_renaming_method_parameters                    63
cancel_subscriptions                                56
```

**Pull request merged**